cli/src/generate/templates/lib.rs

```diff
@@ -4,10 +4,9 @@
 //! tree-sitter [Parser][], and then use the parser to parse some code:
 //!
 //! ```
-//! let code = r#"
-//! "#;
+//! let code = "";
 //! let mut parser = tree_sitter::Parser::new();
-//! parser.set_language(&tree_sitter_PARSER_NAME::language()).expect("Error loading CAMEL_PARSER_NAME grammar");
+//! parser.set_language(&tree_sitter_PARSER_NAME::LANGUAGE.into()).unwrap();
```

Comment:
What do you think about `set_language` taking in `LanguageFn` then?  That would simplify this part to

``` rust
parser.set_language(tree_sitter_PARSER_NAME::LANGUAGE).unwrap();
```

It's just aesthetic really but I like how it reduces the noise a bit.

Reply:
Yeah, it's true the `into()` thing is a bit noisy.

My thing with `set_language` is that now we have a second code path for creating `Language` objects, which doesn't go through a `LanguageFn`, which is loading them from WASM via [`WasmStore::load_language`](https://github.com/tree-sitter/tree-sitter/blob/527848d7c0a8f76962d676293ee8742b3f1e23ee/lib/binding_rust/wasm_language.rs#L58). So I wouldn't want to change the type to take a `LanguageFn`.

Although, we could (I guess) make it take an `impl Into<Language>`...

Reply:
> My thing with `set_language` is that now we have a second code path for creating `Language` objects, which doesn't go through a `LanguageFn`: loading languages from WASM files via [`WasmStore::load_language`](https://github.com/tree-sitter/tree-sitter/blob/527848d7c0a8f76962d676293ee8742b3f1e23ee/lib/binding_rust/wasm_language.rs#L58). So I wouldn't want to change the type to take a `LanguageFn`.

Oh that's nice!  I hadn't been following the wasm stuff closely.  So that's how you're doing dynamic linking if though it's not very well standardized on wasm yet?


> Although, we could (I guess) make it take an `impl Into<Language>`, so that you could pass either a `Language` or a `LanguageFn`.

That would definitely be a nice way to provide backwards-compatibility.  The new `tree-sitter` crate with this change would still work with old language grammar releases, since you'd just have the `Language` directly to pass in.  You can update your grammar dependencies at your leisure and it would be a one-liner for each to change how you're passing the language into your parser.

Also, if you went with `TryInto` instead I think that could even subsume the error handling from `WasmStore::load_language`.  So you'd end up creating a wrapper along the lines of

``` rust
pub struct WasmLanguage<'a> {
  store: &'a mut WasmStore,
  name: &'a str,
  bytes: &'a [u8],
}

impl<'a> TryFrom<WasmLanguage<'a>> for Language {
  type Error = WasmError;
  fn try_from(wasm_lang: WasmLanguage<'a>) -> Result<Result, WasmError> {
    // the body of `WasmStore::load_language`
  }
}

impl WasmStore {
  pub fn language(&mut self, name: &str, bytes: &[u8]) -> WasmLanguage {
    WasmLanguage { store: self, name, bytes }
  }
}

// let store: WasmStore = ...;
// parser.set_language(store.language("python", python_bytes))?;
```

Reply:
I think I'm wrong about my suggestion that `Into` or `TryInto` would provide transparent backwards compatibility, since `set_language` currently takes a _reference_ to a language.  Given that, I retract my suggestions — I think the dependency inversion on its own is a huge win, and worth getting in ASAP.  We can iterate on the `set_language` ergonomics later if it turns out to be an actual pain point.